File: app/src/main/java/com/example/mike/mp3player/client/views/adapters/SearchResultAdapter.java
package com.example.mike.mp3player.client.views.adapters;

import android.support.v4.media.MediaBrowserCompat;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.AlbumArtPainter;
import com.example.mike.mp3player.client.views.viewholders.MyFolderViewHolder;
import com.example.mike.mp3player.client.views.viewholders.MySongViewHolder;
import com.example.mike.mp3player.client.views.viewholders.RootItemViewHolder;
import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.commons.MediaItemUtils;

import javax.inject.Inject;

import static com.example.mike.mp3player.commons.Constants.MEDIA_ITEM_TYPE;

public class SearchResultAdapter extends MediaItemRecyclerViewAdapter {

    @Inject
    public SearchResultAdapter(AlbumArtPainter albumArtPainter) {
        super(albumArtPainter);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = null;
        if (viewType == MediaItemType.SONG.getValue()) {
            // create a new views
            layoutInflater = LayoutInflater.from(parent.getContext());
            ViewGroup v = (ViewGroup) layoutInflater
                    .inflate(R.layout.song_item_menu, parent, false);
            return new MySongViewHolder(v, albumArtPainter);
        }
        else if (viewType == MediaItemType.ROOT.getValue()) {
            // create a new views
            layoutInflater = LayoutInflater.from(parent.getContext());
            ViewGroup v = (ViewGroup) layoutInflater
                    .inflate(R.layout.root_item_menu, parent, false);
            return new RootItemViewHolder(v, albumArtPainter);
        }
        else if (viewType == MediaItemType.FOLDER.getValue()) {
            layoutInflater = LayoutInflater.from(parent.getContext());
            ViewGroup v = (ViewGroup) layoutInflater
                    .inflate(R.layout.folder_item_menu, parent, false);
            return new MyFolderViewHolder(v, albumArtPainter);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        MediaBrowserCompat.MediaItem item = items.get(holder.getAdapterPosition());
        if (null == item) {
            return;
        }
        MediaItemType mediaItemType = MediaItemUtils.getMediaItemType(item);
        if (null != mediaItemType) {
            switch (mediaItemType) {
                case SONG:
                    MySongViewHolder mySongViewHolder = (MySongViewHolder) holder;
                    mySongViewHolder.bindMediaItem(item);
                    break;
                case FOLDER:
                    MyFolderViewHolder folderViewHolder = (MyFolderViewHolder) holder;
                    folderViewHolder.bindMediaItem(item);
                    break;
                case ROOT:
                    RootItemViewHolder rootItemViewHolder = (RootItemViewHolder) holder;
                    rootItemViewHolder.bindMediaItem(item);
                    break;
            } // switch
        } // if MediaItemType not null
    }

    @Override
    public int getItemViewType(int position) {
        MediaBrowserCompat.MediaItem mediaItem = items.get(position);
        MediaItemType mediaItemType = (MediaItemType) MediaItemUtils.getExtra(MEDIA_ITEM_TYPE, mediaItem);
        if (null != mediaItemType) {
            return mediaItemType.getValue();
        }
        return super.getItemViewType(position);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

}

File: client/src/main/java/com/github/goldy1992/mp3player/client/views/adapters/SearchResultAdapter.kt
package com.github.goldy1992.mp3player.client.views.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.viewbinding.ViewBinding
import com.github.goldy1992.mp3player.client.AlbumArtPainter
import com.github.goldy1992.mp3player.client.views.viewholders.MediaItemViewHolder
import com.github.goldy1992.mp3player.client.views.viewholders.MyFolderViewHolder
import com.github.goldy1992.mp3player.client.views.viewholders.MySongViewHolder
import com.github.goldy1992.mp3player.client.views.viewholders.NoSearchResultsViewHolder
import com.github.goldy1992.mp3player.client.views.viewholders.RootItemViewHolder
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.commons.MediaItemUtils
import com.github.goldy1992.mp3player.databinding.FolderItemMenuBinding
import com.github.goldy1992.mp3player.databinding.NoSearchResultsItemMenuBinding
import com.github.goldy1992.mp3player.databinding.RootItemMenuBinding
import com.github.goldy1992.mp3player.databinding.SongItemMenuBinding
import javax.inject.Inject

class SearchResultAdapter

    @Inject
    constructor(albumArtPainter: AlbumArtPainter) : MediaItemRecyclerViewAdapter(albumArtPainter) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MediaItemViewHolder<ViewBinding> {
        Log.e(logTag(), "creating new view holder")
        val context : Context = parent.context
        val layoutInflater: LayoutInflater = LayoutInflater.from(context)
        return when (viewType) {
            MediaItemType.SONG.value -> { // create a new views
                val view = SongItemMenuBinding.inflate(layoutInflater, parent, false)
                MySongViewHolder(context, view, albumArtPainter)
            }
            MediaItemType.ROOT.value -> { // create a new views
                val view = RootItemMenuBinding.inflate(layoutInflater, parent, false)
                RootItemViewHolder(context, view, albumArtPainter)
            }
            MediaItemType.FOLDER.value -> {
                val view = FolderItemMenuBinding.inflate(layoutInflater, parent, false)
                MyFolderViewHolder(context, view, albumArtPainter)
            }
            else -> {
                createEmptyViewHolder(parent)
            }
        }
    }

    override fun onBindViewHolder(holder: MediaItemViewHolder<ViewBinding>, position: Int) {
        val item = getItem(position)
        Log.e(logTag(), "on bind view holder, item: " + item.mediaId)
        holder.bindMediaItem(item)
    }

    override fun getItemViewType(position: Int): Int {
        val mediaItem = getItem(position)
        val mediaItemType : MediaItemType? = MediaItemUtils.getMediaItemType(mediaItem)
        return mediaItemType?.value ?: super.getItemViewType(position)
    }

    override fun createEmptyViewHolder(parent: ViewGroup): NoSearchResultsViewHolder {
        val view = NoSearchResultsItemMenuBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return NoSearchResultsViewHolder(parent.context, view, albumArtPainter)
    }

    /**
     * @return the name of the log tag given to the class
     */
    override fun logTag(): String {
        return "SEARCH_RESULT_ADAPTER"
    }
}